internal/r8/deobfuscator.go

```diff
@@ -0,0 +1,222 @@
+package r8
+
+import (
+	"bufio"
+	"fmt"
+	"log"
+	"os"
+	"regexp"
+	"strings"
+)
+
+type StacktraceMethod struct {
+	reference string
+	callSite  string
+}
+
+type StacktraceType struct {
+	name        string
+	indentation int
+	methods     []StacktraceMethod
+}
+
+type MappedType struct {
+	obfuscated StacktraceType
+	realName   string
+}
+
+type MappedMethodCall struct {
+	reference string
+	key       string
+}
+
+func Deobfuscate(stacktrace *string, mapFilePath string) string {
+	types := findUniqueTypes(stacktrace)
+	mapping := findMappingFor(types, mapFilePath)
+
+	deobfuscated := *stacktrace
+	for k, v := range mapping {
+		deobfuscated = strings.ReplaceAll(deobfuscated, k, v)
+	}
+
+	return deobfuscated
+}
+
+func findUniqueTypes(stacktrace *string) []StacktraceType {
+	symbolPattern := regexp.MustCompile("^\\s*at (.+)(?:\\.(.+)\\((.+)\\))$")
+	sourceFilePattern := regexp.MustCompile("SourceFile:(\\d+)")
+
+	var symbols = make([]*StacktraceType, 0)
+	scanner := bufio.NewScanner(strings.NewReader(*stacktrace))
+
+	for scanner.Scan() {
+		line := scanner.Text()
+		item := symbolPattern.FindStringSubmatchIndex(line)
+		if item != nil {
+			typeIndex := item[2]
+			typeName := line[typeIndex:item[3]]
+			methodName := line[item[4]:item[5]]
+			callSite := line[item[6]:item[7]]
+			sourceFileMatch := sourceFilePattern.FindStringSubmatch(callSite)
+			if sourceFileMatch != nil {
+				methodName = fmt.Sprintf("%s:%s", methodName, sourceFileMatch[1])
+			}
+			symbol := getSymbolForType(typeName, symbols)
+			if symbol == nil {
+				symbol = &StacktraceType{typeName, typeIndex + 1, make([]StacktraceMethod, 0)}
+				symbols = append(symbols, symbol)
+			}
+			method := &StacktraceMethod{methodName, callSite}
+			if !containsMethod(symbol.methods, method) {
+				symbol.methods = append(symbol.methods, *method)
+			}
+		}
+	}
+
+	if err := scanner.Err(); err != nil {
+		log.Fatal(err)
+	}
+
+	var result = make([]StacktraceType, len(symbols))
+	for i, v := range symbols {
+		result[i] = *v
+	}
+
+	return result
+}
+
+func containsMethod(methods []StacktraceMethod, method *StacktraceMethod) bool {
+	for _, v := range methods {
+		if v == *method {
+			return true
+		}
+	}
+
+	return false
+}
+
+func getSymbolForType(typeName string, symbols []*StacktraceType) *StacktraceType {
+	for _, v := range symbols {
+		if v.name == typeName {
+			return v
+		}
+	}
+	return nil
+}
+
+func findMappingFor(symbols []StacktraceType, mapFilePath string) map[string]string {
+	mapFile, err := os.Open(mapFilePath)
+	if err != nil {
+		log.Fatal(err)
```

Comment:
log.Fatal will terminate the program with non-0 exit code. I suggest that we propagate the error to caller instead